pkg/cmd/experimental/diagnostics/diagnostics.go

```diff
@@ -43,141 +71,148 @@ can be used with subcommands.
 `
 
 func NewCommandDiagnostics(name string, fullName string, out io.Writer) *cobra.Command {
-	opts := options.NewAllDiagnosticsOptions(out)
+	o := &OverallDiagnosticsOptions{
+		RequestedDiagnostics: AvailableOverallDiagnostics.List(),
+		LogOptions:           &log.LoggerOptions{Out: out},
+	}
+
 	cmd := &cobra.Command{
 		Use:   name,
 		Short: "This utility helps you understand and troubleshoot OpenShift v3.",
 		Long:  fmt.Sprintf(longAllDescription, fullName),
 		Run: func(c *cobra.Command, args []string) {
-			opts.GlobalFlags = c.PersistentFlags()
-			run.Diagnose(opts)
+			kcmdutil.CheckErr(o.Complete())
+
+			failed, err := o.RunDiagnostics()
+			o.Logger.Summary()
+			o.Logger.Finish()
+
+			kcmdutil.CheckErr(err)
+			if failed {
+				os.Exit(255)
+			}
+
 		},
 	}
 	cmd.SetOutput(out) // for output re: usage / help
-	opts.BindFlags(cmd.Flags(), options.NewAllDiagnosticsFlagInfos())
-	// Although we reuse DiagOptions across all commands, we do not want the flags buried
-	// in the "global" flags, so we add them locally at each command.
-	opts.DiagOptions.BindFlags(cmd.Flags(), options.NewDiagnosticsFlagInfos())
-
-	/*
-	   This command needs the client factory built in the "client" subcommand.
-	   Generating the factory adds flags to the "client" cmd, and we do not want
-	   to add those flags to this command (the only client option here is a config
-	   file). So the factory object from client cmd is reused for this command.
-	*/
-	clientCmd, factory := NewClientCommand("client", name+" client", out)
-	opts.ClientDiagOptions.Factory = factory
-
-	cmd.AddCommand(clientCmd)
-	cmd.AddCommand(NewMasterCommand("master", name+" master", out))
-	cmd.AddCommand(NewNodeCommand("node", name+" node", out))
+
+	o.Factory = osclientcmd.New(cmd.Flags()) // side effect: add standard persistent flags for openshift client
```

Comment:
This adds a crapton of flags in the wrong place. See `openshift ex diagnostics -h` for what I was trying to avoid with the previous circumnavigation. I think a refactor of the factory generation might be a bit of a tall order though.